zetaclient/orchestrator/chain_activate.go

```diff
@@ -0,0 +1,248 @@
+
+	ticker := time.NewTicker(common.ZetaBlockTime * 2)
+	for {
+		select {
+		case <-ticker.C:
+			oc.ActivateAndDeactivateChains()
+		case <-oc.stop:
+			oc.logger.Std.Info().Msg("WatchChainActivation stopped")
+			return
+		}
+	}
+}
+
+// ActivateAndDeactivateChains activates and deactivates chains according to chain params and config file
+//
+// The chains to be activated:
+//   - chain params flag 'IsSupported' is true AND
+//   - chain is configured in config file
+//
+// The chains to be deactivated:
+//   - chain params flag 'IsSupported' is false OR
+//   - chain is not configured in config file
+//
+// Note:
+//   - zetaclient will reload config file periodically and update in-memory config accordingly.
+//   - As an tss signer, please make sure the config file is always well configured and not missing any chain
+func (oc *Orchestrator) ActivateAndDeactivateChains() {
+	// create new signer and observer maps
+	// Note: the keys of the two maps are chain IDs and they are always exactly matched
+	newSignerMap := make(map[int64]interfaces.ChainSigner)
+	newObserverMap := make(map[int64]interfaces.ChainObserver)
+
+	// create new signers and observers
+	oc.CreateSignerObserverEVM(newSignerMap, newObserverMap)
+	oc.CreateSignerObserverBTC(newSignerMap, newObserverMap)
+
+	// activate newly supported chains and deactivate chains that are no longer supported
+	oc.DeactivateChains(newObserverMap)
+	oc.ActivateChains(newSignerMap, newObserverMap)
+}
+
+// DeactivateChains deactivates chains that are no longer supported
+func (oc *Orchestrator) DeactivateChains(
+	newObserverMap map[int64]interfaces.ChainObserver,
+) {
+	// loop through existing observer map to deactivate chains that are not in new observer map
+	oc.mu.Lock()
+	defer oc.mu.Unlock()
+	for chainID, observer := range oc.observerMap {
+		_, found := newObserverMap[chainID]
+		if !found {
+			oc.logger.Std.Info().Msgf("DeactivateChains: deactivating chain %d", chainID)
+			observer.Stop()
+
+			// remove signer and observer from maps
+			delete(oc.signerMap, chainID)
+			delete(oc.observerMap, chainID)
+			oc.logger.Std.Info().Msgf("DeactivateChains: deactivated chain %d", chainID)
+		}
+	}
+}
+
+// ActivateChains activates newly supported chains
+func (oc *Orchestrator) ActivateChains(
+	newSignerMap map[int64]interfaces.ChainSigner,
+	newObserverMap map[int64]interfaces.ChainObserver,
+) {
+	// loop through new observer map to activate chains that are not in existing observer map
+	for chainID, observer := range newObserverMap {
+		_, found := oc.observerMap[chainID]
+		if !found {
+			oc.logger.Std.Info().Msgf("ActivateChains: activating chain %d", chainID)
+
+			// open database and load data
+			err := observer.LoadDB(oc.dbPath)
+			if err != nil {
+				oc.logger.Std.Error().
+					Err(err).
+					Msgf("ActivateChains: error LoadDB for chain %d", chainID)
+				continue
+			}
+			observer.Start()
+
+			// add signer and observer to maps
+			oc.mu.Lock()
```

Comment:
Why not using
```
	oc.mu.Lock()
	defer oc.mu.Unlock()
```
At the beggining of the method

Reply:
Because the `observer.LoadDB(oc.dbPath)` is an IO operation which could potentially hold the lock for longer time.